File: packages/api-core/src/pipeline.ts
import { Middleware, Next } from "./types";
import { Request } from "./request";
import { Context } from "./context";
import { Response } from "./response";

export class Pipeline {
  private middlewares: Middleware[] = [];

  use(mw: Middleware): void {
    this.middlewares.push(mw);
  }

  compose(finalHandler: Next): Next {
    const mws = [...this.middlewares];
    return async (req: Request, ctx: Context): Promise<Response> => {
      let idx = -1;
      const dispatch = async (i: number, r: Request, c: Context): Promise<Response> => {
        if (i <= idx) throw new Error("next() called multiple times");
        idx = i;
        const mw = mws[i];
        if (!mw) return finalHandler(r, c);
        return mw.handle(r, c, (nr, nc) => dispatch(i + 1, nr, nc));
      };
      return dispatch(0, req, ctx);
    };
  }
}

File: packages/api-core-python/src/aptx_api_core/_pipeline.py
from __future__ import annotations

from typing import Awaitable, Callable

from ._context import Context
from ._request import Request
from ._response import Response


class Pipeline:
    def __init__(self) -> None:
        self._middlewares: list = []

    def use(self, mw) -> None:
        self._middlewares.append(mw)

    def compose(
        self, final: Callable[[Request, Context], Awaitable[Response]]
    ) -> Callable[[Request, Context], Awaitable[Response]]:
        mws = list(self._middlewares)

        async def dispatch(i: int, req: Request, ctx: Context) -> Response:
            if i < len(mws):
                async def nxt(r: Request, c: Context) -> Response:
                    return await dispatch(i + 1, r, c)

                return await mws[i].handle(req, ctx, nxt)
            return await final(req, ctx)

        return lambda req, ctx: dispatch(0, req, ctx)
